src/cosalib/metal.py

```diff
@@ -0,0 +1,319 @@
+        self.force = kwargs.get("force", False)
+        self.kargs = kwargs.pop("kargs", [])
+        self.size = kwargs.pop("size", "padded")
+        self.target_args = kwargs.pop("target_args", [])
+
+        OSTreeBuild.__init__(self, *args, **kwargs)
+
+        # Read in the image.yaml file.
+        self.img_cfg = None
+        img_cfg_f = kwargs.get(
+            "image_config",
+            os.path.join(self.workdir, "src", "config", "image.yaml"),
+        )
+        if not os.path.exists(img_cfg_f):
+            raise Exception(f"failed to find file {img_cfg_f}")
+        with open(img_cfg_f, 'r') as f:
+            self.img_cfg = yaml.safe_load(f.read())
+        if not self.img_cfg:
+            raise Exception(f"{img_cfg_f} did not read in properly")
+
+        # this is used in case the image has a different disk
+        # name than the platform
+        self.platform_image_name = kwargs.get(
+            "platform_image_name", self.platform)
+
+        # building dasd for ppc64el make no sense
+        if self.arch != "any" and self.arch != self.basearch:
+            raise Exception(
+                f"{self.basearch} cannot be built for disk {variant}")
+        # only support known-good artitectures, eg. no sparks
+        if self.basearch not in ["x86_64", "aarch64", "s390x", "ppc64le"]:
+            raise Exception(
+                f"{self.basearch} is not support for metal-type disk images")
+
+    @property
+    def image_name(self):
+        if self._image_name is not None:
+            return self._image_name
+
+        return (f'{self.build_name}-{self.build_id}'
+                f'-{self.platform}.{self.basearch}.{self.format}')
+
+    @property
+    def rootfs_size(self):
+        """
+        Returns the rootfs size and the image size.
+        """
+        size = self.ostree_repo_estimated_size
+        return f"{int(size * 1.35)}M"
+
+    @property
+    def disk_size(self):
+        """
+        Returns the size of the disk. This is the size of the
+        rootfs with padding
+        """
+        size = self.img_cfg.get("size")
+        if size:
+            if str(size).upper().endswith("M"):
+                return size
+            else:
+                return f"{size}G"
+        return f"{self.rootfs_size + 513}M"
+
+    @property
+    def rootfs_type(self):
+        rootfs = self.img_cfg.get("rootfs", DEFAULT_ROOTFS)
+        rootfs_luks = self.img_cfg.get("luks_rootfs", "None").lower()
+        if rootfs_luks in ("yes", "1", "true"):
+            rootfs = "luks"
+            log.debug("rootfs will be housed in LUKS")
+        return rootfs
+
+    def get_disk_args(self, ref):
+        disk_args = []
+
+        tmod_cmd = run_verbose(
+            [
+                "ostree", f"--repo={self.ostree_repo}",
+                "ls", ref, "/usr/lib/modules",
+            ],
+            check=True, capture_output=True
+        ).stdout.decode('utf-8').splitlines()[-1]
```

Comment:
Not a blocker, this is OK as is - just noting the OSTree API exposed by gobject-introspection supports this and basically everything else we're doing via a subprocess.  Sometimes the subprocess is less verbose though.

Reply:
BTW for plenty of examples of using the API see https://github.com/ostreedev/ostree-releng-scripts which is actually already a git submodule here (though I don't think we use it).

Reply:
I took a look and even took a look at using them per your comment. I would like to switch this over to using the native libs in the future. The problem with using the `ostree-releng-scripts` is that they are Py2 :face_with_head_bandage:

Reply:
The introspection API itself works in py2 and py3. We actually already use it in a bunch of places in the codebase (you can grep for e.g. `gi.require_version`).

But yeah, definitely not worth blocking on.